### ISSUE TITLE
[RFC] Speed up python function and arg serialization in RPC APIs

### ISSUE THREAD

**Issue by rohan-varma** [pytorch/pytorch#49911]
## 🚀 Feature
In our current RPC APIs (rpc_sync, async, rpc.remote), we first serialize (pickle) the function, argument, and kwargs being sent to the destination worker, and pass this serialized representation as a string in C++ to construct our notion of a `SerializedPyObj`. This is done so that we can send these bytes across the network and reconstruct the needed information on the remote end.

We have noticed some slowness in this python serialization, namely in use cases with a large number of remote trainers being set up as an RRef on a master/parameter server node. In setups like these, the following pattern is quite common:

```
# On parameter server
model = create_model(...)
for trainer_name in trainers:
    trainer_rref = rpc.remote(
        _setup_model_on_trainer,
        model,
        ...additional arguments, expensive to serialize
    )
```

Note that in the loop while the dst worker is different, the function and arguments being serialized are the same. This may mean that caching the serialized representation from call to call could be useful, speeding up the end to end latency. For speeding up individual calls, we can look into parallelizing the serialization, which will be more complex than simple caching as the parallelization would have to be implemented in the C++ layer.

## Motivation
In some training workloads, we have noticed serialization taking up a large portion of time when initializing the setup discussed above. Since many of these workloads follow the pattern above, caching/parallelization should be a good candidate to speed this up. 

## Pitch
Implement caching similar to the demo here: and evaluate its impact on the above described workloads. If parallelization across arguments in individual calls also sounds like it makes sense and is likely to result in a speedup, we can discuss a concrete implementation plan to move the pickling into C++ to take advantage of multiple cores.

**Comment by mrshenli** [comment]
From the demo PR, looks like the args can contain a lot of small tensors. If that's the common case, RDMA in TensorPipe won't be sufficient. cc @lw and @beauby  for this use case. 

A follow-up question is in which layer we should put this caching/serialization logic? Probably not in TensorPipe to avoid over-complicating the comm layer. Two other options are in RPC or layers on top of RPC (e.g., applications, or helpers). Will it work if we provide a new decorator for RPC functions? E.g.:

```python
@rpc.functions.enable_caching
def my_function_to_cache(*args):
    pass
```

**Comment by lw** [comment]
We had some ideas to optimize transfers of small tensors in TensorPipe, for example writing new channels which sacrifice some bandwidth in exchange for a smaller per-tensor overhead (which is likely the dominating factor in small tensors). But we had bigger priorities and didn't get to it.

However from @rohan-varma's description the issue doesn't seem to lie in the transport layer but in the pickling itself, which also adds a constant per-tensor overhead. In that respect, there's nothing we can do in TensorPipe: that's all Python overhead, higher up in the stack.

I thought that the "silver bullet" for all Python woes was to use TorchScript, in which pickling should be considerably faster. Has this been tried for the above demo? What gains did it bring? Could this be our "official" solution?

As for caching, to me that seems a prickly approach. Since basically all Python objects are mutable (damned Python dynamism) we cannot safely reuse a cached serialization unless we first verify that the object hasn't changed, and this is an operation that takes linear time in the object's size. Hence it could be as slow as just pickling straight away.

> move the pickling into C++ to take advantage of multiple cores

Are you sure this is really feasible? I thought pickling requires holding the GIL, which would mean that even if it's done in C++ it cannot be done in parallel by multiple threads. Unless you were talking about TorchScript pickling, but as I said above that should already be faster, without any changes required in RPC, right?

**Comment by rohan-varma** [comment]
@lw True, after some more thought I agree moving pickling into C++ won't likely give us too much since we would need to immediately grab the GIL. TorchScript in general is our recommendation for improved performance and if this code were indeed TorchScripted, it would go through TS pickling, but some use cases may not be able to fully TorchScript their training loop. I think we can probably still proceed with the caching portion of this RFC, though.

**Comment by rohan-varma** [comment]
> From the demo PR, looks like the args can contain a lot of small tensors. If that's the common case, RDMA in TensorPipe won't be sufficient. cc @lw and @beauby for this use case.
> 
> A follow-up question is in which layer we should put this caching/serialization logic? Probably not in TensorPipe to avoid over-complicating the comm layer. Two other options are in RPC or layers on top of RPC (e.g., applications, or helpers). Will it work if we provide a new decorator for RPC functions? E.g.:
> 
> ```python
> @rpc.functions.enable_caching
> def my_function_to_cache(*args):
>     pass
> ```

Sorry for the late reply, but we are actually trying to cache the results of pickling (python) args and functions themselves, not the actual result of the function. The idea is that many trainers issue the same RPCs and do (potentially) expensive pickling, when we could just be returning a cached result. Caching for application functions is tricky due to issues like idempotency and would probably be best handled by the user app (or we can provide the decorator as you mentioned).

**Comment by lw** [comment]
One issue with caching the pickled data is mutability. The application might know that the data won't change between two calls, but PyTorch doesn't and thus it would have to check, which could potentially be as slow as just pickling again.

I wonder if, as a start, we could fix this in the application: the user could _explicitly_ pickle what it wants to send over, obtain `bytes` objects for it, send those `bytes` as arguments of the RPC call, and have the remote function unpickle them back. This might be easier said than done, since if the values contain tensors it could get tricky. 

Maybe we should first try a few variations of these potential solutions in the application, and thus validate their correctness, usability and performance, before considering "upstreaming" the best one into PyTorch?
